### ISSUE TITLE
Compaction not guaranteed to run on successful write to "source" series

### ISSUE THREAD

**Issue by kylebebak** [RedisTimeSeries/RedisTimeSeries#1065]
Disclaimer: the bug described below may have already been fixed. The following behavior was observed running RTS `v1.4.5`, commit `ef4baec` from `2020-09-16`.

-----

We created a "raw" time series and "compacted" time series with the following commands:

```
TS.CREATE raw LABELS agg_type none
TS.CREATE 10m LABELS agg_type 10m
TS.CREATERULE raw 10m AGGREGATION SUM 600000
```

We let a process run for a while that writes entries to the `raw` series. When the process was done, these were the entries in the raw series. The full contents are in this text file: [raw_data_at_end.txt](https://github.com/RedisTimeSeries/RedisTimeSeries/files/8104713/raw_data_at_end.txt)

```
TS.RANGE raw 1645021200000 +

   1) 1) (integer) 1645021200906
      2) 1
   2) 1) (integer) 1645021201983
      2) 1
   ...
```

There are 1496 entries with a total sum of 1498. All entries had a value of "1", except for one entry with a value of "3".

Then we ran some `TS.RANGE` commands to see what was in the `10m` series.

```
127.0.0.1:6380> TS.RANGE 10M - +
1) 1) (integer) 1645021200000
   2) 600
3) 1) (integer) 1645023600000
   2) 300
127.0.0.1:6380>
```

And here are the contents of the `raw` series, using a 10m aggregation period:

```
127.0.0.1:6380> TS.RANGE RAW - + AGGREGATION SUM 600000
1) 1) (integer) 1645021200000
   2) 600
2) 1) (integer) 1645021800000
   2) 401
3) 1) (integer) 1645023600000
   2) 300
4) 1) (integer) 1645024200000
   2) 197
127.0.0.1:6380>
```

**The 10m series is simply missing the `401` count in the `1645021800000` compaction period**. The entries in the adjacent `1645021200000` and `1645023600000` compaction periods are present.

In other words, when running our version of RTS, it's possible for a write to a "source" series to fail to trigger compaction to a "target" series, but the write to the source series still succeeds.

**This is very concerning**. If "partial" writes are possible, as they seem to be based on our observations, this is a critical bug in RTS. For RTS to be dependable as a time series database, the write to the source series plus compaction to all target series needs to be atomic.

**Comment by kylebebak** [comment]
@OfirMos 

All entires were inserted using `TS.MADD <key> * <count> ...`. We use `*` for the [timestamp arg](https://oss.redis.com/redistimeseries/commands/#tsmadd) for each sample we add. So samples are inserted in ascending time order.

Here's a script that inserts the same samples that are in this raw series: 
[add_samples.txt](https://github.com/RedisTimeSeries/RedisTimeSeries/files/8111356/add_samples.txt). Running this script with `redis-cli < add_samples.txt` doesn't reproduce the bug for me. But this script specifies timestamps, and doesn't use `*`, as we did when we observed this bug. Writing a script that uses `*` and also gets the timestamps to match would be nearly impossible.

@gkorland 

We're going to upgrade to the latest RTS build, `v1.6.8`, thank you for suggesting this.

Does it sound possible to either of you that writes to a source series and compaction to its target series don't happen atomically? I didn't see anything in issues or releases suggesting this was a known bug, but I also didn't look very carefully.

**Comment by LiorKogan** [comment]
@kylebebak 

Is there a specific reason you are calling `TS.MADD` and not `TS.ADD`? Do you input more than one sample per call? And if so, do you use '*' more than once in the same `TS.MADD` call?

**Comment by kylebebak** [comment]
@LiorKogan 

We're using `TS.MADD` because in our real use case we're adding samples to multiple streams in a single call. One of these streams is the `raw` stream I mentioned above.

**We never add more than one sample to a given stream in a `TS.MADD` call**. For example, a call would look like this: `TS.MADD raw * <number> raw_1 * <number> raw_2 * <number>`

I only mentioned that we're using `TS.MADD` because I wanted to describe repro conditions as accurately as possible

**Comment by OfirMos** [comment]
@kylebebak thanks for the reply.
Another question, are you using Redis Cluster or is it standalone Redis?

**Comment by kylebebak** [comment]
Thank you @OfirMos and everyone else for looking into this!

We're using a standalone Redis, all the writes are happening on the same Redis. I ran `INFO MEMORY` on the server to make sure writes weren't being dropped because we hit memory limits, but we're nowhere close, which is good

```
used_memory:44242264
used_memory_human:42.19M
used_memory_peak:46959616
used_memory_peak_human:44.78M
total_system_memory:8324198400
total_system_memory_human:7.75G
```

**Comment by kylebebak** [comment]
Another update: when we observed the issue above, we also observed some other issues.

In our real use case, we have some additional time series, the `raw` series doesn't only compact into the `10m` series, it compacts into a `60m` series as well

```
TS.CREATE raw LABELS agg_type none
TS.CREATE 10m LABELS agg_type 10m
TS.CREATE 60m LABELS agg_type 60m
TS.CREATERULE raw 10m AGGREGATION SUM 600000
TS.CREATERULE raw 60m AGGREGATION SUM 3600000
```

`TS.MRANGE` reads the following samples from the `10m` and `60m` series:

From `10m` series:

```
[
  [
    # ...
    [
      1645020000000,
      "600"
    ],
    [
      1645020600000,
      "600"
    ],
    [
      1645021200000,
      "600"
    ],
    [
      1645023600000,
      "300"
    ]
  ]
]
```

From `60m` series

```
[
  [
    # ...
    [
      1645016400000,
      "3600"
    ]
  ]
]
```

We can see that the `10m` series has samples at `1645020000000`, and 1 hour later at `1645023600000`. The `60m` series has a sample at `1645016400000`, **but none at `1645020000000`**.

This is also a bug. The `raw` series is the source for the `10m` and `60m` series, and the raw series has samples with timestamps g.t. `1645024200000`, which is why there's a `1645023600000` sample in the `10m` series.

In other words, the `raw` series has samples with timestamps g.t. `1645020000000 + 1 hour and 10 minutes`, but compaction from the raw series to the `60m` series to create a sample at `1645020000000` wasn't triggered.

-----

Today we're going to run our code with RTS `v1.6.8`, I'll report on whether we still see this issue.

**Comment by kylebebak** [comment]
Another update: we've upgrade from RTS 1.4.5 to 1.6.8 a few days ago and haven't hit any of the above issues yet.

I'm not sure these issues have been resolved in 1.6.8, but at least we haven't hit them.
